src/persist-client/src/internal/apply.rs

```diff
@@ -156,164 +238,301 @@ where
+                }
+                ApplyCmdResult::SkippedStateTransition((seqno, err, maintenance)) => {
+                    return Ok((seqno, Err(err), maintenance));
+                }
+                ApplyCmdResult::Indeterminate(err) => {
+                    return Err(err);
+                }
+                ApplyCmdResult::ExpectationMismatch(seqno) => {
+                    cmd.cas_mismatch.inc();
+                    self.fetch_and_update_state(Some(seqno)).await;
+                }
+            }
+        }
     }
 
     async fn apply_unbatched_cmd_locked<
         R,
         E,
         WorkFn: FnMut(SeqNo, &PersistConfig, &mut StateCollections<T>) -> ControlFlow<E, R>,
     >(
-        state: &mut TypedState<K, V, T, D>,
+        state: &LockingTypedState<K, V, T, D>,
         cmd: &CmdMetrics,
-        mut work_fn: WorkFn,
+        work_fn: &mut WorkFn,
         cfg: &PersistConfig,
         metrics: &Metrics,
         shard_metrics: &ShardMetrics,
         state_versions: &StateVersions,
-    ) -> Result<(SeqNo, Result<R, E>, RoutineMaintenance), Indeterminate> {
-        let is_write = cmd.name == metrics.cmds.compare_and_append.name;
-        let is_rollup = cmd.name == metrics.cmds.add_and_remove_rollups.name;
-        cmd.run_cmd(shard_metrics, |cas_mismatch_metric| async move {
-            let mut garbage_collection;
-            let mut expiry_metrics;
-
-            loop {
-                let was_tombstone_before = state.collections.is_tombstone();
-
-                let (work_ret, mut new_state) = match state.clone_apply(cfg, &mut work_fn) {
-                    Continue(x) => x,
-                    Break(err) => {
-                        return Ok((state.seqno(), Err(err), RoutineMaintenance::default()))
-                    }
-                };
-                expiry_metrics = new_state.expire_at((cfg.now)());
-
-                // Sanity check that all state transitions have special case for
-                // being a tombstone. The ones that do will return a Break and
-                // return out of this method above. The one exception is adding
-                // a rollup, because we want to be able to add a rollup for the
-                // tombstone state.
-                //
-                // TODO: Even better would be to write the rollup in the
-                // tombstone transition so it's a single terminal state
-                // transition, but it'll be tricky to get right.
-                if was_tombstone_before && !is_rollup {
-                    panic!(
-                        "cmd {} unexpectedly tried to commit a new state on a tombstone: {:?}",
-                        cmd.name, state
-                    );
+    ) -> ApplyCmdResult<K, V, T, D, R, E> {
+        let computed_next_state =
+            state.read_lock(&metrics.locks.applier_read_noncacheable, |state| {
+                match Self::compute_next_state_locked(state, work_fn, metrics, cmd, cfg) {
```

Comment:
we just match on this error anyway as soon as we get out of the `state.read_lock` call, so lets remove that match here and do the Err transformation in the match below